### PR TITLE
Add bottom to CupertinoNavigationBar to allow for a double-row nav bar

### LINKED ISSUE #10469: Support segmented controls in nav bars and double row nav bars
Extracting from #7508

Examples:
screen shot 2017-06-02 at 3 59 59 pm

screen shot 2017-06-02 at 4 00 16 pm

### PR DESCRIPTION
**Pull request by victorsanni**
Fixes [Support segmented controls in nav bars and double row nav bars](https://github.com/flutter/flutter/issues/10469)

Screenshot 2024-09-30 at 3 09 04 PM

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: examples/api/lib/cupertino/nav_bar/cupertino_navigation_bar.1.dart [hunk 1]
```diff
@@ -0,0 +1,73 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/cupertino.dart';
+
+/// Flutter code sample for [CupertinoNavigationBar].
```
Comment (MitchellGoodwin) [inline examples/api/lib/cupertino/nav_bar/cupertino_navigation_bar.1.dart:7]: Probably needs a more descriptive explanation.

File: examples/api/lib/cupertino/nav_bar/cupertino_navigation_bar.1.dart [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  State<NavBarExample> createState() => _NavBarExampleState();
+}
+
+class _NavBarExampleState extends State<NavBarExample> {
+  @override
+  Widget build(BuildContext context) {
+    return CupertinoPageScaffold(
+      navigationBar: CupertinoNavigationBar(
+        // Try removing opacity to observe the lack of a blur effect and of sliding content.
+        backgroundColor: CupertinoColors.systemGrey.withOpacity(0.5),
```
Comment (MitchellGoodwin) [inline examples/api/lib/cupertino/nav_bar/cupertino_navigation_bar.1.dart:36]: I don't think this example needs to show this behavior.

File: packages/flutter/lib/src/cupertino/nav_bar.dart [hunk 3]
```diff
@@ -448,6 +449,16 @@ class CupertinoNavigationBar extends StatefulWidget implements ObstructingPrefer
...
 
+  /// A widget to place at the bottom of the navigation bar.
+  ///
+  /// Only widgets that implement [PreferredSizeWidget] can be used at the
+  /// bottom of a navigation bar.
+  ///
+  /// See also:
+  ///
+  ///  * [PreferredSize], which can be used to give an arbitrary widget a preferred size.
+  final PreferredSizeWidget? bottom;
```
Comment (MitchellGoodwin) [inline packages/flutter/lib/src/cupertino/nav_bar.dart:460]: In the doc comments should link to the example.

**Review by MitchellGoodwin — APPROVED** [review]
LGTM!

File: packages/flutter/lib/src/cupertino/nav_bar.dart [hunk 4]
```diff
@@ -458,7 +476,8 @@ class CupertinoNavigationBar extends StatefulWidget implements ObstructingPrefer
 
   @override
   Size get preferredSize {
-    return const Size.fromHeight(_kNavBarPersistentHeight);
+    final double heightForDrawer = bottom != null ? bottom!.preferredSize.height : 0.0;
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/cupertino/nav_bar.dart:479]: `bottom?.preferredSize.height ?? 0.0`

**Review by justinmc — APPROVED** [review]
LGTM :+1:

**Review by LongCatIsLooong — APPROVED** [review]

**Pull request merged**